src/pudl/transform/ferc1.py

```diff
@@ -242,31 +242,42 @@ def drop_duplicate_rows_dbf(df, params: DropDuplicateRowsDbf):
     duplicates are dropped with ``keep="first"``. If any duplicates do not contain the
     same data, an assertion will be raised.
     """
-    if params is None:
-        params = DropDuplicateRowsDbf()
-    if params.data_columns:
-        pks = (
-            pudl.metadata.classes.Package.from_resource_ids()
-            .get_resource(params.table_name.value)
-            .schema.primary_key
-        )
-        # checks to make sure the drop is targeted as expected
-        pk_dupes = df[df.duplicated(pks, keep=False)]
-        pk_data_dupes = pk_dupes[
-            pk_dupes.duplicated(pks + params.data_columns, keep=False)
-        ]
-        dupes_w_unique_data = pk_dupes.loc[
-            pk_dupes.index.difference(pk_data_dupes.index)
-        ]
-        if not dupes_w_unique_data.empty:
-            raise AssertionError(
-                "Duplicates have unique data and should not be dropped. Unique data: "
-                f"{len(dupes_w_unique_data)}: {dupes_w_unique_data}"
+    pks = (
+        pudl.metadata.classes.Package.from_resource_ids()
+        .get_resource(params.table_name.value)
+        .schema.primary_key
+    )
+    # add a column that indicates whether or not any of the data columns contain null data
+    df.loc[:, "null_data"] = df[params.data_columns].isnull().any(axis="columns")
+
+    # checks to make sure the drop is targeted as expected
+    # of the PK dupes, drop all instances when the data *is also the same*
+    dupes_w_unique_data = (
+        df[df.duplicated(pks, keep=False)].drop_duplicates(
+            pks + params.data_columns, keep=False
+        )
+        # if there are pk+data dupes, is there one record with some null data
+        # an other with completely non-null data??
+        .assign(
+            count_of_null_vs_non_null=lambda x: x.groupby(pks)[["null_data"]].transform(
+                "nunique"
             )
-        logger.info(
-            f"Dropping {(len(pk_data_dupes)/2)/len(df):.1%} of duplicate records."
         )
-        df = df.drop_duplicates(pks, keep="first")
+    )
+    if not dupes_w_unique_data[
+        dupes_w_unique_data.count_of_null_vs_non_null != 2
+    ].empty:
+        raise AssertionError(
+            "Duplicates have unique data and should not be dropped. Unique data: "
+            f"{len(dupes_w_unique_data)}: {dupes_w_unique_data}"
```

Comment:
This seems good and will work for the case in my #2134 PR.

However, there are some edge cases that would bypass this assertion error:
- dupes in `dupes_w_unique_data` that both have some NA values (this seems tricky any maybe something we should combine? But highly uncertain about this. Depends on the case.)
- dupes in `dupes_w_unique_data` where one row has some NA values but other non-NA values that don't match the other duplicate (idk what we would even do with this besides flag it...)

Reply:
I'm glad it will probably work with your table!

I'm glad you are noting these edge cases. I believe the first example that you noted will actually raise the assertion. And I think that is a good thing for now. I could definitely imagine developing a transformer or a part of this transformer that would condense these, but I would suggest we cross that bridge when we get there. For now, I added a small unit test  to be sure that these types of records trip the assertion:
```
# if a PK dupes has mismatched null data
df_nulls = pd.read_csv(
    StringIO(
        """
report_year,utility_id_ferc1,asset_type,data_col1,data_col2
2021,71,stuff,,700
2021,71,stuff,70,
"""
    )
)
with pytest.raises(AssertionError):
    drop_duplicate_rows_dbf(df_nulls, params=params)
```

you are very right that this would not raise an error...
```
report_year,utility_id_ferc1,asset_type,data_col1,data_col2
2021,71,stuff,,701
2021,71,stuff,70,700
```
I can definitely imagine this showing up. sigh. I'll try to weasel in a way to make this raise the assertion and add it as a unit test.

Reply:
okay I edited the assertion in there and added a section to the unit test to catch this second edge case

Reply:
It takes me like 5 min every time to wrap my head around all the negatives in 

```
if not dupes_w_unique_data[
        dupes_w_unique_data.count_of_null_vs_non_null != 2
    ].empty
```
lolol